### ISSUE TITLE
Change the Button block's Edit component for reuse in other blocks

### ISSUE THREAD

**Issue by pablinos** [WordPress/gutenberg#20634]
**Is your feature request related to a problem? Please describe.**
Whilst developing [some](https://github.com/Automattic/jetpack/blob/master/extensions/blocks/eventbrite/edit.js#L375) [blocks](https://github.com/Automattic/jetpack/blob/master/extensions/blocks/calendly/edit.js#L166) in Jetpack we have found the need to have a button component for use in the edit component of the blocks that is practically identical to the `core/button`  edit component but without the URL editing controls.

This has been solved multiple times but each time has involved [reimplementing the button edit component](https://github.com/Automattic/jetpack/blob/master/extensions/shared/submit-button.js) i.e. copying over the code and adjusting it.

I spoke to @mtias about this issue and he suggested looking at block variations as a solution. I think this may still work, but would still need some tweaks to the core block's code.

**Describe the solution you'd like**
I would like to export the edit component from `core/button` (so it can be imported from the `@wordpress/block-library` package) and add a prop that would allow the URL editing controls to be disabled.

**Describe alternatives you've considered**
As mentioned above I have investigated using block variations for this, and there is the current solution of copying the code over. If we didn't want to export the component then I could probably gain access to it as the block is registered or by grabbing it from the block registry, but I would still need to make the change to disable the URL editing UI

**Comment by mtias** [comment]
> I spoke to @mtias about this issue and he suggested looking at block variations as a solution. I think this may still work, but would still need some tweaks to the core block's code.

This is how I would expect it to be architected. Exposing `edit` interfaces is not something I think we should do. It's a very heavy-handed and opaque coupling, since we cannot ensure "edit" won't change in ways that you didn't expect. We can commit to handling the lack of an attribute in a certain way, since that's something we can easily build tests for.

We either use variations (and handle some common cases in the core block) or we export the atomic parts of the block interface.

**Comment by pablinos** [comment]
Thanks @mtias. It sounds like using block variations would be the best route for us, for now, so I'll go ahead with that and see where it takes us.

> Exposing edit interfaces is not something I think we should do

I know we don't do this explicitly, currently, but we have exposed all the settings of the blocks via a filter as the blocks are registered, and they can be retrieved after they're registered. Because this is to allow people to extend a block, it's likely that there will be some reliance on these components in the wild.

> We either use variations (and handle some common cases in the core block) 

This can be done using variations and an attribute (or two) passed through to the inner block. It doesn't strike me as the cleanest implementation, and would probably mean that we need to do more checks on the integrity of the post content. I also see some disadvantages with not being able to change the variation. All this aside, I'll work up a proof of concept so we can see how it looks and whether it's in line with what you're thinking.

> we export the atomic parts of the block interface.

I think with the number of hooks that are being developed for things like colour settings and borders, this is also a viable option but doesn't solve the problem of having an alternative implementation. As changes are made to the core button they would have to be added to the version used in these embed blocks too, which is the situation we have at the moment.

**Comment by mtias** [comment]
> As changes are made to the core button they would have to be added to the version used in these embed blocks too, which is the situation we have at the moment.

Right, but that's also exactly why exposing edit interfaces is not great, because now updating the block is not just a matter of handling its own internal implementation but also the fact it could be consumed in multiple ways by other blocks we have no idea about.

**Comment by pablinos** [comment]
> it could be consumed in multiple ways by other blocks we have no idea about.

Isn't that already the case? We allow people to extend blocks by gaining access to their settings (and therefore the edit components), so they are likely to be used in those ways already. 

I'm wondering whether buttons are a special case, as we have a button component and a button block. I can't think of other blocks like that.
